glide-core/src/client/value_conversion.rs

```diff
@@ -442,6 +443,89 @@ pub(crate) fn convert_to_expected_type(
+           6) (integer) 24529
+        3) "engines"
+        4) 1) "LUA"
+           2) 1) "libraries_count"
+              2) (integer) 3
+              3) "functions_count"
+              4) (integer) 5
+
+        1) "running_script"
+        2) (nil)
+        3) "engines"
+        4) ...
+
+        RESP3 response example
+        1# "running_script" =>
+           1# "name" => "<function name>"
+           2# "command" =>
+              1) "fcall"
+              2) "<function name>"
+              ... rest `fcall` args ...
+           3# "duration_ms" => (integer) 5000
+        2# "engines" =>
+           1# "LUA" =>
+              1# "libraries_count" => (integer) 3
+              2# "functions_count" => (integer) 5
+        */
+        // First part of the response (`running_script`) is converted as `Map[str, any]`
+        // Second part is converted as `Map[str, Map[str, int]]`
+        ExpectedReturnType::FunctionStatsReturnType => match value {
+            // TODO reuse https://github.com/Bit-Quill/glide-for-redis/pull/331 and https://github.com/aws/glide-for-redis/pull/1489
+            Value::Map(map) => {
+                if map[0].0 == Value::BulkString(b"running_script".into()) {
+                    // already a RESP3 response - do nothing
+                    Ok(Value::Map(map))
+                } else {
+                    // cluster (multi-node) response - go recursive
+                    convert_map_entries(
+                        map,
+                        Some(ExpectedReturnType::BulkString),
+                        Some(ExpectedReturnType::FunctionStatsReturnType),
+                    )
+                }
+            }
+            Value::Array(array) if array.len() == 4 => {
+                let mut result: Vec<(Value, Value)> = Vec::with_capacity(2);
+                let running_script_info = array[1].clone();
+                let running_script_converted = match running_script_info {
+                    Value::Nil => Ok(Value::Nil),
+                    Value::Array(inner_map_as_array) => {
+                        convert_array_to_map_by_type(inner_map_as_array, None, None)
+                    }
+                    _ => Err((ErrorKind::TypeError, "Response couldn't be converted").into()),
+                };
+                result.push((array[0].clone(), running_script_converted?));
+
+                let Value::Array(engines_info) = array[3].clone() else {
+                    return Err((ErrorKind::TypeError, "Incorrect value type received").into());
+                };
+
+                let engines_info_converted = convert_array_to_map_by_type(
+                    engines_info,
+                    Some(ExpectedReturnType::BulkString),
+                    Some(ExpectedReturnType::Map {
+                        key_type: &None,
+                        value_type: &None,
+                    }),
+                );
+                result.push((array[2].clone(), engines_info_converted?));
```

Comment:
This is a lot of clones. Can they be avoided by doing something like this?
```suggestion
            Value::Array(mut array) if array.len() == 4 => {
                let mut result: Vec<(Value, Value)> = Vec::with_capacity(2);
                let running_script_info = array.remove(1);
                let running_script_converted = match running_script_info {
                    Value::Nil => Ok(Value::Nil),
                    Value::Array(inner_map_as_array) => {
                        convert_array_to_map_by_type(inner_map_as_array, None, None)
                    }
                    _ => Err((ErrorKind::TypeError, "Response couldn't be converted").into()),
                };
                result.push((array.remove(0), running_script_converted?));

                let Value::Array(engines_info) = array.remove(3) else {
                    return Err((ErrorKind::TypeError, "Incorrect value type received").into());
                };

                let engines_info_converted = convert_array_to_map_by_type(
                    engines_info,
                    Some(ExpectedReturnType::BulkString),
                    Some(ExpectedReturnType::Map {
                        key_type: &None,
                        value_type: &None,
                    }),
                );
                result.push((array.remove(2), engines_info_converted?));
```

Reply:
Thanks, that worked